react/ComponentStyle.md

```diff
@@ -0,0 +1,304 @@
+
+Madadata.propTypes = {
+  name: PropTypes.string.isRequired,
+};
+
+// don't use context if you don't know what it is.
+// Madadata.contextTypes = {}
+
+export default Madadata;
+```
+
+* signature
+
+```
+(props, context) => JSXElement;
+```
+
+* Notes
+  * capitalize the **first character** of your component.
+  * optional params of this function can be **props**, **context**.
+  * specify the propTypes of your component for props and context **explicitly**.
+  * export your component as **default**.
+  * name the file as **ComponentName.jsx**.
+
+* Styles
+  * prefer arrow function
+  * If JSX spans for several lines
+    ```
+    const Madadata = props => (
+      <div>
+        {props.name}
+      </div>
+    );
+    ```
+  * prefer collapsing props in propTypes and contextTypes and prefer isRequired to enforce checking prop existence.
+    ```
+    Madadata.propTypes = {
+      name: PropTypes.string.isRequired,
+    };
+
+    Madadata.contextTypes = {
+      store: PropTypes.object.isRequired,
+    }
+    ```
+
+# PureComponent
+
+* when to use?
+  * has lifecycle method
+  * has state to keep.
+  * UI is determined only by state, props.
+    * UI do not depend on global variable.
+    * UI do not use any utility function that has variables from outside.
+    * UI do not have children that are not PureComponent.
+
+* code snippet
+
+```
+// PureMadadata.jsx
+import React, { PureComponent, PropTypes } from 'react';
+
+class PureMadadata extends PureComponent {
+
+  constructor(props, context) {
+    super(props, context);
+    this.state = { value: 'madadata' };
+    this.handleSomethingClick = this.handleSomethingClick.bind(this);
+  }
+
+  componentWillMount() {
+    // * call this.setState here to update state, the next render will pick up state changes.
+    // * will only be invoked once.
+  }
+
+  componentDidMount() {
+    // * do setTimeout, setInterval, addEventListeners
+    // * do ajax calls (dispatch actions that will be handled by redux-saga)
+    // * access any refs to your children
+  }
+
+  componentWillReceiveProps(nextProps) {
+    // * componentWillReceiveProps enables you to decide what to do with next props before the component is rendered.
+    // * this.props -> old props
+    // * nextProps -> new props
+    // * will not be called for initial render
+    // * call this.setState here to update state, the next render will pick up state changes.
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    // * it will always return true by default. This method enables you to check to see whether you want react to re-render this component with new props and state. It's a costly operation and also a performance optimization point.
+    // * PureComponent will help you **shallow** compare the this.props and nextProps, this.state and nextState. If you need to do compare deeply, explicitly do this in the shouldComponentUpdate method of a Component instead of PureComponent.
```

Comment:
so if we use PureComponent, the component won't render if there's change at a deep level?